Store silicon laws on brain + law prediction

ScarKy0 opened the pull request:
<!-- Guidelines: https://docs.spacestation14.io/en/getting-started/pr-guideline -->

## About the PR
<!-- What did you change? -->
Majority of things borg law related have been moved to shared.
Borgs laws are now stored on the brain, instead of on the chassis.

~~Requires #43020~~ Merged

## Why / Balance
<!-- Discuss how this would affect game balance or explain why it was changed. Link any relevant discussions or issues. -->
Prediction good.
Linking borg laws to the brain is also a TODO on the silicon design document.

What we aim to achieve here is make players not want to gib borgs over simple ion storms, to do this we tie the laws to the brain (instead of chassis) and add a way to reset them (in another PR).

## Technical details
<!-- Summary of code changes for easier review. -->
SiliconLawBoundComponent has been given the ability to store cached laws and whether it is subverted.
To properly achieve the effect of laws on brain there have been some changes:
- SiliconLawProvider now gets linked  to a SiliconLawBound. SiliconLawBound fetches it's provider when initialized via the until now unused GetSiliconLawsEvent.
- The UI and everything seeing law related is now based on the values cached in SiliconLawBound. This is to ensure admins see the same lawset as the players in cases of potential crysis (like laws not syncing properly)
- We cache the values in SiliconLawBound for prediction purposes. This has to be done, as to not allow third party clients to read silicon laws while avoiding any messy networking.

The basic flow of the changes should look like this:
1. A brain gets initialized, it runs FetchLawset() to find a provider for its laws. Since it itself is a provider, if fetches its own lawset.
2. Once the brain is inserted into a chassis, the chassis LawBound gets linked to the SiliconLawProvider in the brain, meaning any laws the chassis reads will be actually from the brain.
3. When the brain gets removed from the chassis, we unlink them. This causes the laws of the chassis to remain unchanged but no longer receive updates from the brain.

For cases when we must take priority of the chassis laws (for example, XenoBorgs have their own lawset once inserted into the chassis) we prioritize using SiliconLawProvider from the owning entity before checking anything else, meaning if a brain gets inserted into a chassis with SiliconLawProvider, the chassis will remain linked to its own laws instead of the brain.

## Media
<!-- Attach media if the PR makes in-game changes (clothing, items, features, etc).
Small fixes/refactors are exempt. Media may be used in SS14 progress reports with credit. -->

https://github.com/user-attachments/assets/e039b193-c525-40a1-8534-8d78a3f1122e

https://github.com/user-attachments/assets/df3629e5-6fa6-4bc8-b744-a4a30d4715db

## Requirements
<!-- Confirm the following by placing an X in the brackets [X]: -->
- [X] I have read and am following the [Pull Request and Changelog Guidelines](https://docs.spacestation14.com/en/general-development/codebase-info/pull-request-guidelines.html).
- [X] I have added media to this PR or it does not require an in-game showcase.
<!-- You should understand that not following the above may get your PR closed at maintainer’s discretion -->

## Breaking changes
<!-- List any breaking changes, including namespaces, public class/method/field changes, prototype renames; and provide instructions for fixing them.
This will be posted in #codebase-changes. -->
Majority of methods in serverside `SiliconLawSystem` have been moved to shared and had their signature changed to `Entity<T>` instead of `(uid, comp)`
`SiliconLawsSaveMessage` now takes a new `OverrideProvider` parameter, which indicates whether the provider of the laws should be modified or the entity itself.
`SiliconLawBuiState` has been removed. Updating the UI is now done via `AfterAutoHandleState` in the clientside `SiliconLawSystem`.
`SiliconLawSystem.GetLaws` has been split into `GetBoundLaws` and `GetProviderLaws`. Former getting the laws cached in a `SiliconLawBoundComponent` and latter getting the laws directly from a `SiliconLawProviderComponent`. The methods are now also in `SharedSiliconLawSystem`.
`IonStormSystem.IonStormTarget()` now takes `Entity<SiliconLawProviderComponent>` instead of `Entity<SiliconLawBoundComponent>`.
`SiliconLawBoundComponent.LastLawProvider` has been removed. Use `LawsetProvider` instead. Do note it gets set to null if not linked to an entity.
`SiliconLawProviderComponent` and `SiliconLawBoundComponent` now only get networked to their owners.
`BaseBorgChassisNotIonStormable` has been removed. Use a chassis with `PositronicBrainNotIonStormable` or `MMINotIonStomable` instead.
`RadioSystem` has been given a shared variant `SharedRadioSystem` with new API for `IntrinsicRadioTransmitterComponent`. The API must be used instead of modifying the component directly.

**Changelog**
<!-- Add a Changelog entry to make players aware of new features or changes that could affect gameplay.
Make sure to read the guidelines and take this Changelog template out of the comment block in order for it to show up.
Changelog must have a :cl: symbol, so the bot recognizes the changes and adds them to the game's changelog. -->
:cl:
- tweak: Silicon law systems have received a major refactor. Please report issues with your laws on the discord or github.
- tweak: Silicon laws are now stored on the brain instead of the chassis.
- tweak: Specific silicon chassis (xenoborgs, syndieborgs, etc) can now override the laws of a brain. This is indicated via a text in the borg panel.
- remove: You can no longer emag cyborg chassis without brains inside.
- add: You can now emag MMIs and Positronic Brains.
<!--
:cl:
- add: Added fun!
- remove: Removed fun!
- tweak: Changed fun!
- fix: Fixed fun!
-->

ScarKy0 committed:
updating from upstream master

ScarKy0 committed:
add back BUI state

ScarKy0 committed:
i hate rider sometimes

ScarKy0 committed:
cleaning up

ScarKy0 committed:
remove useless BUI state

Pok27 commented on Content.Shared/Silicons/Laws/SharedSiliconLawSystem.cs:
```diff
@@ -15,64 +23,273 @@ public abstract partial class SharedSiliconLawSystem : EntitySystem
+
+    private void OnPlayerSpawnComplete(Entity<SiliconLawBoundComponent> ent, ref PlayerSpawnCompleteEvent args)
+    {
+        ent.Comp.LastLawProvider = args.Station;
```
Needs dirtying, when you change the field of a network component

ScarKy0 replied:
Good catch. Honestly I'm not entirely sure it needs it, but might as well dirty it anyways, I believe SiliconLawBoundComp is on the chopping block in the future either way.

ScarKy0 committed:
ugh its so dirty in here

Pok27 reviewed (approved):
LGTM

slarticodefast commented on Content.Client/Silicons/Laws/Ui/SiliconLawMenu.xaml.cs:
```diff
@@ -8,20 +9,27 @@ namespace Content.Client.Silicons.Laws.Ui;
     {
-        state.Laws.Sort();
+        var laws = _lawSystem.GetLaws(uid);
+        var radioComp = _entity.GetComponent<IntrinsicRadioTransmitterComponent>(uid);
```
Make this a `TryGetComponent` in case the entity does not have the `IntrinsicRadioTransmitterComponent`.
The list of channels should remain empty in that case, but the borg should still be able to state laws by speaking.

slarticodefast commented on Content.Shared/Radio/Components/IntrinsicRadioTransmitterComponent.cs:
```diff
@@ -8,9 +8,9 @@ namespace Content.Shared.Radio.Components;
 ///     radio headset).
 /// </summary>
-[RegisterComponent, NetworkedComponent]
+[RegisterComponent, NetworkedComponent, AutoGenerateComponentState]
```
Make this component access restricted to the system and add methods for adding, removing and setting the channels. That will enforce that other systems changing the datafield will always dirty it properly.

slarticodefast commented on Content.Shared/Silicons/Laws/Components/SiliconLawBoundComponent.cs:
```diff
@@ -9,13 +9,14 @@ namespace Content.Shared.Silicons.Laws.Components;
     /// </summary>
-    [DataField]
+    [DataField, AutoNetworkedField]
     public EntityUid? LastLawProvider;
```
Won't this cause problems if the law provider entity is ever deleted? Both because we don't have WER and because the borg would suddenly be without laws. Same if the law provider was outside PVS range on the client.
The whole code here seems kinda weirdly designed to me, why not directly store the current laws instead? In many cases borgs are their own law provider already, so I don't get why we have one component referencing the other one like this.

ScarKy0 replied:
I can not network this one since it's not important for predicted laws but I REALLY don't wanna have to refactor this component out of existance :godo:

slarticodefast commented on Content.Shared/Silicons/Laws/Components/SiliconLawProviderComponent.cs:
```diff
@@ -1,38 +1,41 @@
     /// </summary>
-    [DataField, ViewVariables(VVAccess.ReadWrite)]
+    [DataField, AutoNetworkedField]
     public SiliconLawset? Lawset;
```
Needs `IRobustClonable` implemented if we autonetwork this.

ScarKy0 committed:
Revert "ugh its so dirty in here"

This reverts commit a3ed8a929e931acf4e0ad8e76f522258ff11b4cc.

ScarKy0 committed:
partial review

ScarKy0 committed:
further review

ScarKy0 committed:
silly me

ScarKy0 committed:
so very silly of me

ScarKy0 committed:
cloneable

ScarKy0 commented on Content.Client/Silicons/Laws/SiliconLawSystem.cs:
```diff
@@ -1,9 +1,27 @@
+        SubscribeLocalEvent<SiliconLawProviderComponent, AfterAutoHandleStateEvent>(AfterAutoHandleState);
+    }
+
+    private void AfterAutoHandleState(Entity<SiliconLawProviderComponent> ent, ref AfterAutoHandleStateEvent args)
```
Currently such cases do not exist. Only uses of SiliconLawBound exist only on the borgs themselves.

ScarKy0 committed:
i really hate rider

ScarKy0 committed:
fix law ordering

ScarKy0 committed:
doc

ScarKy0 commented:
I gotta update breaking changes again, but ill do it after review as to not have to constantly rewrite it.

ScarKy0 committed:
handle check

ScarKy0 committed:
initial laws on brain, it compiles, only partially works tho

TODO:
- Make admin law UI able to modify non-providers
- Fix ion storms, make them per-brain and non-provider chassis
- Fix law update sound
- Properly test syndieborgs and xenoborgs
- Posibrains with different roundstart lawsets for xeno and syndieborgs
- Fix emag, it needs to relay to borg brain OR chassis if the chassis itself is the provider for itself
- Rethink chassis providers, show that the chassis itself provides laws instead of the brain if that is the case.

ScarKy0 commented:
also todo: mutual cleanup

ScarKy0 committed:
mutual cleanup

ScarKy0 committed:
seperate files

ScarKy0 committed:
admin ui fix

Velken committed:
mmis and positronic brains

ScarKy0 committed:
hopefully potentially maybe fixed emag

Mot2332 commented:
Watch out with the emag interaction, emagged borgs currently encourage a lot of toxic behavior with all associated. With emag laws being tied to brains, and no mechanical way to identify emmaged borgs, this might encourage round removing cyborgs and putting their brain in the incinerator / vault / space. Particularly bad if innocent cyborgs are involved.  If there was some way to set a cyborgs laws / reset them / at least see them, this might not be so potentially bad.

ScarKy0 committed:
emag cleanup

ScarKy0 committed:
provider warning

TODO:
- Fix emagging, again

ScarKy0 committed:
extra space

Velken committed:
ion storm

Velken committed:
SelfProvider toggle on LawProvider comp added/removed

ScarKy0 committed:
emag finalized

ScarKy0 committed:
clean

ScarKy0 committed:
cleanup

ScarKy0 committed:
cleanup

Velken committed:
derelict fix

ScarKy0 committed:
docs + logging

ScarKy0 committed:
fileld

ScarKy0 committed:
improve admin logging

ScarKy0 committed:
clean not dirty

ScarKy0 committed:
small bugfix

ScarKy0 committed:
Update SharedSiliconLawSystem.Provider.cs

Velken committed:
tests, part 1

Velken committed:
fixed conflict

ScarKy0 commented:
> So how is this gonna work with emags?

You emag the brain or a chassis with a brain inside.

Velken committed:
emag + test for emag

Velken committed:
mmi yaml cleanup

Tayrtahn commented on Content.Shared/Radio/EntitySystems/SharedRadioSystem.cs:
Public API methods? Xmldocs!

ScarKy0 committed:
HOPEFULLY fix mind roles

Will need to seperate the mind changes into a new PR

ScarKy0 committed:
init

ScarKy0 committed:
stuff

ScarKy0 committed:
to

ScarKy0 committed:
i guess

ScarKy0 committed:
BeforeEvents

ScarKy0 committed:
comment

ScarKy0 committed:
meh

ScarKy0 committed:
Merge branch 'add-previous-entity-to-mind-events' of https://github.com/ScarKy0/space-station-14 into bord-law-prediction

# Conflicts:
#	Content.Server/Mind/MindSystem.cs
#	Content.Shared/Mind/Components/MindContainerComponent.cs

ScarKy0 committed:
fix

ScarKy0 committed:
Merge remote-tracking branch 'upstream/master' into bord-law-prediction

ScarKy0 committed:
Merge remote-tracking branch 'upstream/master' into bord-law-prediction

ScarKy0 committed:
predict popup awaiting another upstream merge

ScarKy0 committed:
Merge branch 'master' of https://github.com/space-wizards/space-station-14 into bord-law-prediction

# Conflicts:
#	Content.Server/Mind/MindSystem.cs

ScarKy0 committed:
Merge branch 'master' of https://github.com/space-wizards/space-station-14 into bord-law-prediction

# Conflicts:
#	Content.Shared/Mind/Components/MindContainerComponent.cs

ScarKy0 committed:
DOCUMENTATION!!!!!

Velken committed:
3 more test coverage

ScarKy0 committed:
Merge remote-tracking branch 'upstream/master' into bord-law-prediction

ScarKy0 committed:
suffix for providers

ScarKy0 committed:
docs

ScarKy0 committed:
Merge remote-tracking branch 'upstream/master' into bord-law-prediction

Velken committed:
Merge conflict

Velken committed:
git moment

Velken committed:
OVER 9000

Velken committed:
address tests review

Velken committed:
yaml liner hates us all
